source/dem/dem.cc

```diff
@@ -105,6 +105,38 @@ DEMSolver<dim>::reinitialize_force(
     }
 }
 
+template <int dim>
+std::shared_ptr<Insertion<dim>>
+DEMSolver<dim>::set_insertion_type(const DEMSolverParameters<dim> &parameters)
+{
+  if (parameters.model_parmeters.insertion_method ==
+      Parameters::Lagrangian::ModelParameters::InsertionMethod::uniform)
+    {
+      insertion_object = std::make_shared<UniformInsertion<dim>>(
+        parameters,
+        inserted_this_step,
+        number_of_particles_x_direction,
+        number_of_particles_y_direction,
+        number_of_particles_z_direction);
+    }
+  else if (parameters.model_parmeters.insertion_method ==
+           Parameters::Lagrangian::ModelParameters::InsertionMethod::
+             non_uniform)
+    {
+      insertion_object = std::make_shared<NonUniformInsertion<dim>>(
+        parameters,
+        inserted_this_step,
+        number_of_particles_x_direction,
+        number_of_particles_y_direction,
+        number_of_particles_z_direction);
+    }
+  else
+    {
+      throw "The chosen insertion method is invalid";
+    }
+  return insertion_object;
+}
```

Comment:
I don't understand why instead of passing number_of_x etc. you don't pass directly an insertion parameter structure.